pallets/did/src/lib.rs

```diff
@@ -1140,9 +1152,139 @@ pub mod pallet {
+		/// is required to also be a valid account with enough balance to pay
+		/// for fees.
+		///
+		/// The dispatch origin must be a KILT account with enough funds to
+		/// execute the extrinsic and must correspond to the required DID
+		/// Verification Key.
+		///
+		/// Emits `DidCallDispatched`.
+		#[allow(clippy::boxed_local)]
+		#[pallet::call_index(15)]
+		#[pallet::weight({
+			let dispatch_info = call.get_dispatch_info();
+
+			(<T as pallet::Config>::WeightInfo::dispatch_as().saturating_add(dispatch_info.weight), dispatch_info.class)
+		})]
+		pub fn dispatch_as(
+			origin: OriginFor<T>,
+			did_identifier: DidIdentifierOf<T>,
+			call: Box<DidCallableOf<T>>,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+
+			let verification_key_relationship =
+				call.derive_verification_key_relationship().map_err(Error::<T>::from)?;
+
+			Pallet::<T>::verify_account_authorization(&did_identifier, &who, verification_key_relationship)
+				.map_err(Error::<T>::from)?;
+
+			log::debug!("Dispatch call from DID {:?}", did_identifier);
+
+			#[cfg(not(feature = "runtime-benchmarks"))]
+			let result = call.dispatch(
+				DidRawOrigin {
+					id: did_identifier.clone(),
+					submitter: who,
+				}
+				.into(),
+			);
+			#[cfg(feature = "runtime-benchmarks")]
+			let result = call.dispatch(RawOrigin::Signed(did_identifier.clone()).into());
+
+			let dispatch_event_payload = result.map(|_| ()).map_err(|e| e.error);
+
+			Self::deposit_event(Event::DidCallDispatched(did_identifier, dispatch_event_payload));
+
+			result
+		}
+
+		/// Store a new DID on chain.
+		///
+		/// The DID identifier is derived from the account ID that submits this
+		/// call. The authentication key must correspond to the account ID that
+		/// submitted this call. For accounts that use the ed25519 and sr25519
+		/// schema, the authentication key must be of the
+		/// `DidVerificationKey::Ed25519` or `DidVerificationKey::Sr25519`
+		/// variant and contains the public key. For Ecdsa accounts, the
+		/// `DidVerificationKey::Ecdsa` variant is calculated by hashing the
+		/// Ecdsa public key.
+		///
+		/// If this call is dispatched by an account id that doesn't correspond
+		/// to a public private key pair, the `DidVerificationKey::Account`
+		/// variant shall be used (Multisig, Pure Proxy, Governance origins).
+		/// The resulting DID can NOT be used for signing data and is therefore
+		/// limited to onchain activities.
+		///
+		/// There must be no DID information stored on chain under the same DID
+		/// identifier. This call will fail if there exists a DID with the same
+		/// identifier or if a DID with the same identifier existed and was
+		/// deleted.
+		///
+		/// The origin for this account must be funded and provide the required
+		/// deposit and fee.
+		///
+		/// Emits `DidCreated`.
+		#[pallet::call_index(16)]
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::create_from_account())]
+		pub fn create_from_account(
```

Comment:
I would say that maybe this is the only function that should be trait bounded to cases where `AccountId` can be coalesced into a `DidIdentifier`, right? The bound should not apply to the generic pallet itself.

Reply:
The bounds are required for the `into_account` function. from there they will propagate up to where the generic is actually specified (or set, i.e. in the runtime).
We could add the where clause to both `create_from_account` and `dispatch_as` but since these will be made into enums and are heavily changed by macros, that doesn't work. So we need to keep the where-clause where it is (at least I didn't find a quick way to move it away from where it is). I also think it's fine to have it there, since it's more obvious that this pallet requires special things. If it's for the individual call, it's harder to spot.